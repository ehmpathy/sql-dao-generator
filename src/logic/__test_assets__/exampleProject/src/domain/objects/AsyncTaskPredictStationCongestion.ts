import { DomainEntity } from 'domain-objects';
import { AsyncTask, AsyncTaskStatus } from 'simple-async-tasks';

/**
 * an async task which predicts station congestion by comparing expected arrival time with estimated arrival time
 */
export interface AsyncTaskPredictStationCongestion extends AsyncTask {
  id?: number;
  uuid?: string;
  createdAt?: Date;
  updatedAt?: Date;
  status: AsyncTaskStatus;

  /**
   * the station to run this prediction for
   */
  stationUuid: string;

  /**
   * the event to run the prediction on
   *
   * note
   * - this has a nested dependency on a train, which we use to test sql-schema-control order
   */
  trainLocatedEventUuid: string;
}
export class AsyncTaskPredictStationCongestion
  extends DomainEntity<AsyncTaskPredictStationCongestion>
  implements AsyncTaskPredictStationCongestion
{
  public static unique = ['stationUuid', 'trainLocatedEventUuid'];
  public static updatable = ['status'];
}
