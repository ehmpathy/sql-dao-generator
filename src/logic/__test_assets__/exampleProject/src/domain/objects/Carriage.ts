import { DomainEntity } from 'domain-objects';

export enum CarriagePurpose {
  PASSENGER = 'PASSENGER',
  FREIGHT = 'FREIGHT',
}

/**
 * a carriage carries things or people in a train
 */
export interface Carriage {
  id?: number;
  uuid?: string;
  cin: string; // human readable identifier for carriage
  carries: CarriagePurpose;
  capacity: number; // for passenger carriages, number of passengers; for freight, the volume
}
export class Carriage extends DomainEntity<Carriage> implements Carriage {
  public static primary = ['uuid'] as const; // there is nothing that naturally identifies a carriage -> it is unique only on the unique identifier tracked in our db
  public static unique = ['uuid'] as const; // there is nothing that naturally identifies a carriage -> it is unique only on the unique identifier tracked in our db
  public static updatable = [];
}
