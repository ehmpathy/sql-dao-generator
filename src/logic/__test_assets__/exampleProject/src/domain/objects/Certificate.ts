import { DomainValueObject } from 'domain-objects';

export enum CertificateType {
  LOCOMOTIVE_DRIVING = 'LOCOMOTIVE_DRIVING',
  ENGINE_MAINTENANCE = 'ENGINE_MAINTENANCE',
  CARRIAGE_MAINTENANCE = 'CARRIAGE_MAINTENANCE',
}

export interface Certificate {
  id?: number;
  type: CertificateType;
  industryId: string;
}
export class Certificate
  extends DomainValueObject<Certificate>
  implements Certificate {}
