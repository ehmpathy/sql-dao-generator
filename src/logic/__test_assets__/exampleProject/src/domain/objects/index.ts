export * from './AsyncTaskPredictStationCongestion'
export * from './Carriage';
export * from './Certificate';
export * from './Engineer';
export * from './Geocode';
export * from './Train';
export * from './Locomotive';
export * from './TrainLocatedEvent';
export * from './Station';
export * from './Invoice';
export * from './InvoiceLineItem';
export * from './Price';
