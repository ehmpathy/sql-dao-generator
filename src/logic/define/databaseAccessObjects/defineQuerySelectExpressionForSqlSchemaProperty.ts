import { camelCase, snakeCase } from 'change-case';
import { DomainObjectPropertyMetadata } from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';

export const defineQuerySelectExpressionForSqlSchemaProperty = ({
  sqlSchemaName,
  sqlSchemaProperty,
  domainObjectProperty,
  allSqlSchemaRelationships,
}: {
  sqlSchemaName: string;
  sqlSchemaProperty: SqlSchemaPropertyMetadata;
  domainObjectProperty: DomainObjectPropertyMetadata;
  allSqlSchemaRelationships: SqlSchemaToDomainObjectRelationship[];
}): string => {
  // simple case: property is represented in the db simply as a column
  if (!sqlSchemaProperty.isArray && !sqlSchemaProperty.reference) return `${sqlSchemaName}.${sqlSchemaProperty.name}`;

  // since sql-schema-generator does not support arrays of non-references, we can guarantee that its either a reference or an array of references now
  if (!sqlSchemaProperty.reference) throw new Error('unexpected code path, not supported. this is probably a bug'); // fail fast if our expectation is not met though

  // since we know its a reference, lookup the referenced sqlSchemaRelationship
  const referencedSqlSchemaRelationship = allSqlSchemaRelationships.find(
    (rel) => rel.name.domainObject === sqlSchemaProperty.reference!.of.name,
  );
  if (!referencedSqlSchemaRelationship)
    throw new Error('could not find referenced sql schema relationship. this is a bug within sql-dao-generator'); // fail fast, this should not occur
  const referencedSqlSchemaName = referencedSqlSchemaRelationship.name.sqlSchema;

  // check that its not a DIRECT_BY_NESTING reference to a domain object (one we know will be a domain-value-object) which has its own references (references in references) - we dont support that yet
  if (
    sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING &&
    referencedSqlSchemaRelationship.properties.some(({ sqlSchema: sqlSchemaProperty }) => sqlSchemaProperty.reference)
  ) {
    throw new Error(
      `generating a dao is not supported for domain-value-objects that reference other domain-value-objects yet. '${referencedSqlSchemaName}', referenced by '${camelCase(
        sqlSchemaName,
      )}.${
        domainObjectProperty.name
      }', does not meet this criteria. if this is a valid use case, please submit a ticket.`,
    ); //  nesting domain-value-objects within domain-value-objects is probably edge case enough to not worry about for mvp
  }

  // single reference case: grab the uuid or properties by id
  if (!sqlSchemaProperty.isArray) {
    // if it references by implicit uuid, then we just need to grab the uuid and say its "as" the column
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID) {
      return `
    (
      SELECT ${referencedSqlSchemaName}.uuid
      FROM ${referencedSqlSchemaName} WHERE ${referencedSqlSchemaName}.id = ${sqlSchemaName}.${sqlSchemaProperty.name}
    ) AS ${snakeCase(domainObjectProperty.name)}
        `.trim();
    }

    // if it references it by direct nesting, then we need to get all the props of that object so it can by hydrated on instantiation
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) {
      return `
    (
      SELECT json_build_object(
        'id', ${referencedSqlSchemaName}.id,
        'uuid', ${referencedSqlSchemaName}.uuid,
        ${referencedSqlSchemaRelationship.properties
          .map(
            ({ sqlSchema: referencedSqlSchemaProperty }) =>
              `'${referencedSqlSchemaProperty.name}', ${referencedSqlSchemaName}.${referencedSqlSchemaProperty.name}`,
          )
          .join(',\n        ')}
      ) AS json_build_object
      FROM ${referencedSqlSchemaName} WHERE ${referencedSqlSchemaName}.id = ${sqlSchemaName}.${sqlSchemaProperty.name}
    ) AS ${snakeCase(domainObjectProperty.name)}
        `.trim();
    }
  }

  // array reference case: return a json array of uuids or objects
  if (sqlSchemaProperty.isArray) {
    // if it references by implicit uuid, then we just need to grab the uuids and say its "as" the column
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID) {
      return `
    (
      SELECT COALESCE(array_agg(${referencedSqlSchemaName}.uuid ORDER BY ${referencedSqlSchemaName}_ref.array_order_index), array[]::uuid[]) AS array_agg
      FROM ${referencedSqlSchemaName}
      JOIN unnest(${sqlSchemaName}.${sqlSchemaProperty.name}) WITH ORDINALITY
        AS ${referencedSqlSchemaName}_ref (id, array_order_index)
        ON ${referencedSqlSchemaName}.id = ${referencedSqlSchemaName}_ref.id
    ) AS ${snakeCase(domainObjectProperty.name)}
      `.trim();
    }

    // if it references by direct nesting, then we need to get an array of the full objects
    if (sqlSchemaProperty.reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) {
      return `
    (
      SELECT COALESCE(
        json_agg(
          json_build_object(
            'id', ${referencedSqlSchemaName}.id,
            'uuid', ${referencedSqlSchemaName}.uuid,
            ${referencedSqlSchemaRelationship.properties
              .map(
                ({ sqlSchema: referencedSqlSchemaProperty }) =>
                  `'${referencedSqlSchemaProperty.name}', ${referencedSqlSchemaName}.${referencedSqlSchemaProperty.name}`,
              )
              .join(',\n            ')}
          )
          ORDER BY ${referencedSqlSchemaName}_ref.array_order_index
        ),
        array[]::json[]
      ) AS json_agg
      FROM ${referencedSqlSchemaName}
      JOIN unnest(${sqlSchemaName}.${sqlSchemaProperty.name}) WITH ORDINALITY
        AS ${referencedSqlSchemaName}_ref (id, array_order_index)
        ON ${referencedSqlSchemaName}.id = ${referencedSqlSchemaName}_ref.id
    ) AS ${snakeCase(domainObjectProperty.name)}
      `.trim();
    }
  }

  // fail fast if we reach here, not expected
  throw new Error('unexpected code path. this a bug');
};
