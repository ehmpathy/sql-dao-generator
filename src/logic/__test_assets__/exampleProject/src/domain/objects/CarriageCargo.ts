import { DomainEntity, Ref } from 'domain-objects';
import { Carriage } from './Carriage';

/**
 * .what = cargo which has been allocated to a carriage at a given point in time
 */
export interface CarriageCargo {
  id?: number;
  uuid?: string;

  /**
   * the itinerary during which it is allocated
   */
  itineraryUuid: string; // todo: ref to an itinerary once we declare it

  /**
   * the carriage it has been allocated to
   */
  carriageRef: Ref<typeof Carriage>;

  /**
   * the slot that it has been allocated to on the carriage
   */
  slot: number;

  /**
   * the cargo that has been allocated
   */
  cargoExid: null | string;
}
export class CarriageCargo extends DomainEntity<CarriageCargo> implements CarriageCargo {
  public static primary = ['uuid'] as const;
  public static unique = ['itineraryUuid', 'carriageRef', 'slot'] as const;
  public static updatable = ['cargoExid'];
}
