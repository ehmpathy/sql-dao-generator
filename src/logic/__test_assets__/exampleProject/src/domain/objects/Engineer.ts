import { DomainEntity } from 'domain-objects';
import { Certificate } from './Certificate';

/**
 * a train engineer drives and maintains the train
 */
export interface TrainEngineer {
  id?: number;
  uuid?: string;
  socialSecurityNumberHash: string; // the hash of their social security number
  certificates: Certificate[]; // an array of what certificates they've earned
  name: string;
}
export class TrainEngineer extends DomainEntity<TrainEngineer> implements TrainEngineer {
  public static unique = ['socialSecurityNumberHash'];
  public static updatable = ['name', 'certificates'];
}
