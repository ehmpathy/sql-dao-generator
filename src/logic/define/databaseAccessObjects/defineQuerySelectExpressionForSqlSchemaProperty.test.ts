import {
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { defineQuerySelectExpressionForSqlSchemaProperty } from './defineQuerySelectExpressionForSqlSchemaProperty';

describe('defineQuerySelectExpressionForSqlSchemaProperty', () => {
  it('should define the select expression correctly for a standard column-lookup property', () => {
    const expression = defineQuerySelectExpressionForSqlSchemaProperty({
      sqlSchemaName: 'train_engineer',
      sqlSchemaProperty: {
        name: 'social_security_number_hash',
        isArray: false,
        isNullable: false,
        isUpdatable: false,
        isDatabaseGenerated: false,
        reference: null,
      },
      domainObjectProperty: {
        name: 'socialSecurityNumberHash',
        type: DomainObjectPropertyType.STRING,
      },
      allSqlSchemaRelationships: [], // not needed for this one
    });
    expect(expression).toEqual('train_engineer.social_security_number_hash');
  });
  describe('reference: implicit_by_uuid', () => {
    it('should define the select expression correctly for a solo IMPLICIT_BY_UUID reference', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'train',
        sqlSchemaProperty: {
          name: 'lead_engineer_id',
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.IMPLICIT_BY_UUID,
            of: {
              name: 'TrainEngineer',
              extends: DomainObjectVariant.DOMAIN_ENTITY,
            },
          },
        },
        domainObjectProperty: {
          name: 'leadEngineerUuid',
          type: DomainObjectPropertyType.STRING,
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: {
              domainObject: 'TrainEngineer',
              sqlSchema: 'train_engineer',
            },
            properties: [],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ], // not needed for this one
      });
      expect(expression).toContain('SELECT train_engineer.uuid'); // should select the uuid
      expect(expression).toContain('FROM train_engineer'); // from the right table
      expect(expression).toContain(
        'WHERE train_engineer.id = train.lead_engineer_id',
      ); // filtered on the right id
      expect(expression).toContain('AS lead_engineer_uuid'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
    it('should define the select expression correctly for an array of IMPLICIT_BY_UUID references', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'train',
        sqlSchemaProperty: {
          name: 'assigned_engineer_ids',
          isArray: true,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.IMPLICIT_BY_UUID,
            of: {
              name: 'TrainEngineer',
              extends: DomainObjectVariant.DOMAIN_ENTITY,
            },
          },
        },
        domainObjectProperty: {
          name: 'assignedEngineerUuids',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.STRING,
          },
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: {
              domainObject: 'TrainEngineer',
              sqlSchema: 'train_engineer',
            },
            properties: [],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ], // not needed for this one
      });
      expect(expression).toContain(
        'SELECT COALESCE(array_agg(train_engineer.uuid',
      ); // should select an array of uuids, with coalesce to treat empty array
      expect(expression).toContain(
        'array_agg(train_engineer.uuid ORDER BY train_engineer_ref.array_order_index)',
      ); // should select them by persisted array order
      expect(expression).toContain('FROM train_engineer'); // from the right table
      expect(expression).toContain(
        'JOIN unnest(train.assigned_engineer_ids) WITH ORDINALITY',
      ); // using 'unnest' for performance gains,  using "ordinality" so that we can preserve persisted array_order_index
      expect(expression).toContain(
        'train_engineer_ref (id, array_order_index)',
      ); // name the unnested w/ ordinality table well
      expect(expression).toContain(
        'ON train_engineer.id = train_engineer_ref.id',
      ); // filtered on the right id
      expect(expression).toContain('AS assigned_engineer_uuids'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
  });
  describe('reference: direct_by_nesting', () => {
    it('should define the select expression correctly for a solo DIRECT_BY_NESTING reference', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'train_located_event',
        sqlSchemaProperty: {
          name: 'geocode_id',
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        domainObjectProperty: {
          name: 'geocode',
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'Geocode',
            extends: DomainObjectVariant.DOMAIN_LITERAL,
          },
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: { domainObject: 'Geocode', sqlSchema: 'geocode' },
            properties: [
              {
                domainObject: {
                  name: 'latitude',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'latitude',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'longitude',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'longitude',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ],
      });
      expect(expression).toContain('SELECT json_build_object'); // should grab a json object
      expect(expression).toContain('FROM geocode'); // from the right table
      expect(expression).toContain(
        'WHERE geocode.id = train_located_event.geocode_id',
      ); // filtered on the right id
      expect(expression).toContain('AS geocode'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
    it('should define the select expression correctly for an array of DIRECT_BY_NESTING reference', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'train_located_event',
        sqlSchemaProperty: {
          name: 'geocode_ids',
          isArray: true,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        domainObjectProperty: {
          name: 'geocodes',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: { domainObject: 'Geocode', sqlSchema: 'geocode' },
            properties: [
              {
                domainObject: {
                  name: 'latitude',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'latitude',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'longitude',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'longitude',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ],
      });
      expect(expression).toContain('SELECT COALESCE('); // with coalesce to treat empty array
      expect(expression).toContain('json_build_object('); // should grab a json object
      expect(expression).toContain('FROM geocode'); // from the right table
      expect(expression).toContain(
        'JOIN unnest(train_located_event.geocode_ids) WITH ORDINALITY',
      ); // using 'unnest' for performance gains,  using "ordinality" so that we can preserve persisted array_order_index
      expect(expression).toContain('geocode_ref (id, array_order_index)'); // name the unnested w/ ordinality table well
      expect(expression).toContain('ON geocode.id = geocode_ref.id'); // filtered on the right id
      expect(expression).toContain('AS geocodes'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
    it('should define the select expression correctly for a solo DIRECT_BY_NESTING reference with its own solo DIRECT_BY_NESTING reference', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'invoice',
        sqlSchemaProperty: {
          name: 'line_item_id',
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
            of: {
              name: 'InvoiceLineItem',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        domainObjectProperty: {
          name: 'lineItem',
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'InvoiceLineItem',
            extends: DomainObjectVariant.DOMAIN_LITERAL,
          },
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: {
              domainObject: 'InvoiceLineItem',
              sqlSchema: 'invoice_line_item',
            },
            properties: [
              {
                domainObject: {
                  name: 'id',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'id',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'price',
                  type: DomainObjectPropertyType.REFERENCE,
                  of: {
                    name: 'Price',
                    extends: DomainObjectVariant.DOMAIN_LITERAL,
                  },
                },
                sqlSchema: {
                  name: 'price_id',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: {
                    method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
                    of: {
                      name: 'Price',
                      extends: DomainObjectVariant.DOMAIN_LITERAL,
                    },
                  },
                },
              },
              {
                domainObject: {
                  name: 'title',
                  type: DomainObjectPropertyType.STRING,
                },
                sqlSchema: {
                  name: 'title',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
          new SqlSchemaToDomainObjectRelationship({
            name: { domainObject: 'Price', sqlSchema: 'price' },
            properties: [
              {
                domainObject: {
                  name: 'amount',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'amount',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'currency',
                  type: DomainObjectPropertyType.ENUM,
                  of: ['USD'],
                },
                sqlSchema: {
                  name: 'currency',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ],
      });
      expect(expression).toContain('SELECT json_build_object'); // should grab a json object
      expect(expression).toContain(
        'FROM invoice_line_item WHERE invoice_line_item.id = invoice.line_item_id',
      ); // from the right table
      expect(expression).toContain(
        'FROM price WHERE price.id = invoice_line_item.price',
      ); // from the right table
      expect(expression).toContain('AS line_item'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
    it('should define the select expression correctly for a solo DIRECT_BY_NESTING reference with its own array DIRECT_BY_NESTING reference', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'invoice',
        sqlSchemaProperty: {
          name: 'line_item_id',
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
            of: {
              name: 'InvoiceLineItem',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        domainObjectProperty: {
          name: 'lineItem',
          type: DomainObjectPropertyType.REFERENCE,
          of: {
            name: 'InvoiceLineItem',
            extends: DomainObjectVariant.DOMAIN_LITERAL,
          },
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: {
              domainObject: 'InvoiceLineItem',
              sqlSchema: 'invoice_line_item',
            },
            properties: [
              {
                domainObject: {
                  name: 'id',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'id',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'prices',
                  type: DomainObjectPropertyType.ARRAY,
                  of: {
                    type: DomainObjectPropertyType.REFERENCE,
                    of: {
                      name: 'Price',
                      extends: DomainObjectVariant.DOMAIN_LITERAL,
                    },
                  },
                },
                sqlSchema: {
                  name: 'price_ids',
                  isArray: true,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: {
                    method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
                    of: {
                      name: 'Price',
                      extends: DomainObjectVariant.DOMAIN_LITERAL,
                    },
                  },
                },
              },
              {
                domainObject: {
                  name: 'title',
                  type: DomainObjectPropertyType.STRING,
                },
                sqlSchema: {
                  name: 'title',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
          new SqlSchemaToDomainObjectRelationship({
            name: { domainObject: 'Price', sqlSchema: 'price' },
            properties: [
              {
                domainObject: {
                  name: 'amount',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'amount',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'currency',
                  type: DomainObjectPropertyType.ENUM,
                  of: ['USD'],
                },
                sqlSchema: {
                  name: 'currency',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ],
      });
      expect(expression).toContain('SELECT json_build_object'); // should grab a json object
      expect(expression).toContain(
        'FROM invoice_line_item WHERE invoice_line_item.id = invoice.line_item_id',
      ); // from the right table
      expect(expression).toContain('FROM price'); // from the right table
      expect(expression).toContain('ON price.id = price_ref.id'); // from the right table
      expect(expression).toContain('AS line_item'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
    it('should define the select expression correctly for an array DIRECT_BY_NESTING reference with its own solo DIRECT_BY_NESTING reference', () => {
      const expression = defineQuerySelectExpressionForSqlSchemaProperty({
        sqlSchemaName: 'invoice',
        sqlSchemaProperty: {
          name: 'line_item_ids',
          isArray: true,
          isNullable: false,
          isUpdatable: false,
          isDatabaseGenerated: false,
          reference: {
            method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
            of: {
              name: 'InvoiceLineItem',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        domainObjectProperty: {
          name: 'lineItems',
          type: DomainObjectPropertyType.ARRAY,
          of: {
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'InvoiceLineItem',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
        },
        allSqlSchemaRelationships: [
          new SqlSchemaToDomainObjectRelationship({
            name: {
              domainObject: 'InvoiceLineItem',
              sqlSchema: 'invoice_line_item',
            },
            properties: [
              {
                domainObject: {
                  name: 'id',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'id',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'price',
                  type: DomainObjectPropertyType.REFERENCE,
                  of: {
                    name: 'Price',
                    extends: DomainObjectVariant.DOMAIN_LITERAL,
                  },
                },
                sqlSchema: {
                  name: 'price_id',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: {
                    method: SqlSchemaReferenceMethod.DIRECT_BY_NESTING,
                    of: {
                      name: 'Price',
                      extends: DomainObjectVariant.DOMAIN_LITERAL,
                    },
                  },
                },
              },
              {
                domainObject: {
                  name: 'title',
                  type: DomainObjectPropertyType.STRING,
                },
                sqlSchema: {
                  name: 'title',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
          new SqlSchemaToDomainObjectRelationship({
            name: { domainObject: 'Price', sqlSchema: 'price' },
            properties: [
              {
                domainObject: {
                  name: 'amount',
                  type: DomainObjectPropertyType.NUMBER,
                },
                sqlSchema: {
                  name: 'amount',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
              {
                domainObject: {
                  name: 'currency',
                  type: DomainObjectPropertyType.ENUM,
                  of: ['USD'],
                },
                sqlSchema: {
                  name: 'currency',
                  isArray: false,
                  isNullable: false,
                  isUpdatable: false,
                  isDatabaseGenerated: false,
                  reference: null,
                },
              },
            ],
            decorations: { unique: { sqlSchema: null, domainObject: null } },
          }),
        ],
      });
      expect(expression).toContain('SELECT json_build_object'); // should grab a json object
      expect(expression).toContain('FROM invoice_line_item'); // from the right table
      expect(expression).toContain(
        'ON invoice_line_item.id = invoice_line_item_ref.id',
      );
      expect(expression).toContain(
        'FROM price WHERE price.id = invoice_line_item.price',
      ); // from the right table
      expect(expression).toContain('AS line_item'); // with the correct output name
      expect(expression).toMatchSnapshot();
    });
  });
});
