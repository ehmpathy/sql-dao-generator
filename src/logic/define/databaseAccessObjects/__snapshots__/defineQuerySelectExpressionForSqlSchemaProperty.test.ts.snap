// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineQuerySelectExpressionForSqlSchemaProperty should define the select expression correctly for a solo DIRECT_BY_NESTING reference 1`] = `
"(
      SELECT json_build_object(
        'id', geocode.id,
        'uuid', geocode.uuid,
        'latitude', geocode.latitude,
        'longitude', geocode.longitude
      ) AS json_build_object
      FROM geocode WHERE geocode.id = train_located_event.geocode_id
    ) AS geocode"
`;

exports[`defineQuerySelectExpressionForSqlSchemaProperty should define the select expression correctly for a solo IMPLICIT_BY_UUID reference 1`] = `
"(
      SELECT train_engineer.uuid
      FROM train_engineer WHERE train_engineer.id = train.lead_engineer_id
    ) AS lead_engineer_uuid"
`;

exports[`defineQuerySelectExpressionForSqlSchemaProperty should define the select expression correctly for an array of DIRECT_BY_NESTING reference 1`] = `
"(
      SELECT COALESCE(
        json_agg(
          json_build_object(
            'id', geocode.id,
            'uuid', geocode.uuid,
            'latitude', geocode.latitude,
            'longitude', geocode.longitude
          )
          ORDER BY geocode_ref.array_order_index
        ),
        array[]::json[]
      ) AS json_agg
      FROM geocode
      JOIN unnest(train_located_event.geocode_ids) WITH ORDINALITY
        AS geocode_ref (id, array_order_index)
        ON geocode.id = geocode_ref.id
    ) AS geocodes"
`;

exports[`defineQuerySelectExpressionForSqlSchemaProperty should define the select expression correctly for an array of IMPLICIT_BY_UUID references 1`] = `
"(
      SELECT COALESCE(array_agg(train_engineer.uuid ORDER BY train_engineer_ref.array_order_index), array[]::uuid[]) AS array_agg
      FROM train_engineer
      JOIN unnest(train.assigned_engineer_ids) WITH ORDINALITY
        AS train_engineer_ref (id, array_order_index)
        ON train_engineer.id = train_engineer_ref.id
    ) AS assigned_engineer_uuids"
`;
